package com.learning.journalApp.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.learning.journalApp.entity.User;
import com.learning.journalApp.service.UserDetailsServiceImpl;
import com.learning.journalApp.service.UserService;
import com.learning.journalApp.utils.JwtUtil;

import lombok.extern.slf4j.Slf4j;

@RestController
@RequestMapping("/public")
@Slf4j
public class PublicController {

	@Autowired
	private UserService userService;

	@Autowired
	private AuthenticationManager authenticationManager;
	
	@Autowired
	private UserDetailsServiceImpl userDetailsServiceImpl;
	
	@Autowired
	private JwtUtil jwtUtil;

	@GetMapping("/health-check")
	public String healthCheck() {
		return "OK";
	}

	@PostMapping("/signup")
	public void signup(@RequestBody User user) {
		userService.saveNewUser(user);
	}

	@PostMapping("/login")
	public ResponseEntity<String>  login(@RequestBody User user) {
		try {
			Authentication authenticate = authenticationManager
					.authenticate(new UsernamePasswordAuthenticationToken(user.getUserName(), user.getPassword()));

			UserDetails userDetails=userDetailsServiceImpl.loadUserByUsername(user.getUserName());
			String jwt=jwtUtil.generateToken(userDetails.getUsername());
			return new ResponseEntity<>(jwt,HttpStatus.OK);
		} catch (Exception e) {
			log.error("Exception occurred while createAuthenticationToken ",e);
			return new ResponseEntity<>("Incorrect username or password", HttpStatus.BAD_REQUEST);
 
		}

	}

}
